### PR TITLE
CCDB fetching support in DPL

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
Very barebone, for now, but it works.

**Comment by ktf** [comment]
@chiarazampolli @shahor02 @pzhristov @matthiasrichter @mkrzewic @sawenzel @costing 

This is the initial incarnation. I will factor out the curl code once I add support for etag and non HTTP sources (e.g. a file on XRootD).

It probably makes sense to move the whole caching functionality in a separate device, using unmanaged region shared memory, rather than have every device doing the fetches itself.

We'll need to discuss how to specify / map to the mnemonics that people want to have for those objects (do we really need them?) to actual origin / description / subspec of the messages in O2.

**Comment by ktf** [comment]
@dberzano @davidrohr 

```C++
[ 70%] Building CXX object Detectors/ITSMFT/test/HitAnalysis/CMakeFiles/HitAnalysis.dir/G__HitAnalysisDict.cxx.o
CMakeFiles/O2TPCCAGPUTracking.dir/GlobalTracker/AliGPUReconstruction.cxx.o: In function `std::unique_ptr<ali_tpc_common::tpc_fast_transformation::TPCFastTransform, std::default_delete<ali_tpc_common::tpc_fast_transformation::TPCFastTransform> > AliGPUReconstruction::ReadFlatObjectFromFile<ali_tpc_common::tpc_fast_transformation::TPCFastTransform>(char const*)':
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-1/sources/TPCCAGPUTracking/GlobalTracker/AliGPUReconstruction.cxx:343: undefined reference to `ali_tpc_common::tpc_fast_transformation::TPCFastTransform::TPCFastTransform()'
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-1/sources/TPCCAGPUTracking/GlobalTracker/AliGPUReconstruction.cxx:350: undefined reference to `ali_tpc_common::tpc_fast_transformation::TPCFastTransform::setActualBufferAddress(char*)'
CMakeFiles/O2TPCCAGPUTracking.dir/GlobalTracker/AliGPUReconstructionConvert.cxx.o: In function `AliGPUReconstructionConvert::ConvertNativeToClusterData(ClusterNativeAccessExt*, std::unique_ptr<AliHLTTPCCAClusterData::Data [], std::default_delete<AliHLTTPCCAClusterData::Data []> >*, unsigned int*, ali_tpc_common::tpc_fast_transformation::TPCFastTransform const*, int)':
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-1/sources/TPCCAGPUTracking/../TPCFastTransformation/TPCDistortionIRS.h:225: undefined reference to `ali_tpc_common::tpc_fast_transformation::TPCDistortionIRS::getSpline(int, int) const'
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-1/sources/TPCCAGPUTracking/../TPCFastTransformation/TPCDistortionIRS.h:226: undefined reference to `ali_tpc_common::tpc_fast_transformation::TPCDistortionIRS::getSplineData(int, int) const'
collect2: error: ld returned 1 exit status
```

any idea of what is going wrong?

**Comment by davidrohr** [comment]
@ktf: No idea, in particular since you didn't change anything related, and the alitpccommon tag was not modified since several weeks. This should work.

Anyhow, since the integration into O2 is basically ready #AliceO2Group/AliceO2#1747, perhaps I wouldn't spend time debugging this?

**Comment by matthiasrichter** [comment]
There are some strange things going on with the AliTPCCommon code since Friday, and only for the o2 and o2checkcode setups. On Friday, those were failing because of the a failing unit test from the auxiliary repository. And there was really a change which could have caused this.

Can we debug this on the CI machines? Its always suspicious if there is such a strange behavior  and in most, most cases there is a real problem behind. @dberzano I could give it a try if I get access to the CI machines.

**Comment by matthiasrichter** [comment]
@ktf, thats great, I will have a closer look and try it with the TPC reco workflow.

**Comment by davidrohr** [comment]
@matthiasrichter @ktf : Could this be somehow related to incremental builds.
I.e. does my PR AliceO2Group/AliceO2#1747 and this one run incremental builds on the build directory?
Perhaps they are disturbing each other since I changed the AliTPCCommon folder in the CMake in my PR.

**Comment by ktf** [comment]
Errors unrelated. Merging.

**Comment by sawenzel** [comment]
We seem to have real build problems: ```/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-1/sources/TPCCAGPUTracking/GlobalTracker/AliGPUReconstruction.cxx:343: undefined reference to `ali_tpc_common::tpc_fast_transformation::TPCFastTransform::TPCFastTransform()'
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliTPCCommon/alitpccommon-v2.3.2.1-```

Is anyone working on them? @davidrohr ?

**Comment by davidrohr** [comment]
@sawenzel : As I wrote above, I do not understand why this fails. It shouldn't, and nothing has been changed with the code for quite some time. I could only imagine a side-effect with incremental builds since I changed the PATHs in my PR to integrate AliTPCCommon AliceO2Group/AliceO2#1747, if they share the same infrastructure.

**Comment by sawenzel** [comment]
Sorry. Didn't read the full history. But then I think we should fix this on the build system before merging PRs. @dberzano: Could you take a look into this?

**Comment by Barthelemy** [comment]
Hi @ktf 
There is a CCDB api as described here : https://github.com/AliceO2Group/AliceO2/tree/dev/CCDB
Could you use it ? Maybe you can port your code inside the CCDB library/API so that it benefits others ?

**Comment by ktf** [comment]
Yes, I am aware of it. I did not want to use it in the current iterations for a few reasons, most notably:

* It only supports ROOT payloads.
* It predates the polymorphic allocators work of @mkrzewic so it does not write directly in the message payloads.
* most importantly I wanted to get a grip of what DPL needed from curl, before actually going and modify the current one.

That said, I am sure we can discuss how to adapt the current incarnation to something which are all happy with. Besides the first few points, I have a few more things I would also like to add:

* Support the E-Tag mechanism to avoid refetching payloads which are already known.
* Add support for `curl_multi_*` API so that we are not limited to only one concurrent connection.
* I would also like to move the actual fetching to a separate device, so that we reduce by a factor 100 the CCDB connections when no local cache is present.

**Comment by mkrzewic** [comment]
+1 on having this functionality in a separate device at some point, offering acces via shared mem, then it does not matter where it lives. Although I agree with previous comment by @Barthelemy that everybody would benefit from common functionality to not be buried inside DPL...

**Comment by davidrohr** [comment]
+1 for both of mikolaj's comments
Kind Regards
David Rohr

(Sent from my mobile, excuse the typos.)

**Comment by ktf** [comment]
@davidrohr @mkrzewic maybe I should have been more explicit, but yes, that's the plan.

File: Framework/Core/src/WorkflowHelpers.cxx [hunk 1]
```diff
@@ -184,6 +184,7 @@ void WorkflowHelpers::injectServiceDevices(WorkflowSpec& workflow)
       switch (input.lifetime) {
         case Lifetime::Timer: {
           auto concrete = DataSpecUtils::asConcreteDataMatcher(input);
+          processor.options.push_back(ConfigParamSpec{ "period-" + input.binding, VariantType::Int, 1000, { "period of the timer" } });
```
Comment (knopers8) [inline Framework/Core/src/WorkflowHelpers.cxx]: @ktf 
How do I specify in the code the period of timer after this fix? Putting the same option inside DataProcessorSpec with desired period results in throwing 'option is ambiguous and matches different versions of 'period-timer-publish'

Comment (ktf) [reply Framework/Core/src/WorkflowHelpers.cxx]: I probably should avoid pushing the option if it's already there... :-/

**Pull request merged**